build/lib/src/state/filesystem_cache.dart

```diff
@@ -16,144 +14,233 @@ abstract interface class FilesystemCache {
 }
 
 /// [FilesystemCache] that stores data in memory.
 class InMemoryFilesystemCache implements FilesystemCache {
   /// Cached results of [readAsBytes].
-  final _bytesContentCache = LruCache<AssetId, Uint8List>(
-    1024 * 1024,
-    1024 * 1024 * 512,
-    (value) => value.lengthInBytes,
-  );
-
-  /// Pending [readAsBytes] operations.
-  final _pendingBytesContentCache = <AssetId, Future<Uint8List>>{};
+  final _bytesContentCache = <AssetId, Uint8List>{};
 
   /// Cached results of [exists].
   ///
   /// Don't bother using an LRU cache for this since it's just booleans.
-  final _canReadCache = <AssetId, Future<bool>>{};
+  final _canReadCache = <AssetId, bool>{};
 
   /// Cached results of [readAsString].
   ///
   /// These are computed and stored lazily using [readAsBytes].
   ///
   /// Only files read with [utf8] encoding (the default) will ever be cached.
-  final _stringContentCache = LruCache<AssetId, String>(
-    1024 * 1024,
-    1024 * 1024 * 512,
-    (value) => value.length,
-  );
+  final _stringContentCache = <AssetId, String>{};
 
-  /// Pending `readAsString` operations.
-  final _pendingStringContentCache = <AssetId, Future<String>>{};
+  final _pendingWrites = <AssetId, void Function()>{};
+  final _pendingDeletes = <AssetId, void Function()>{};
 
   @override
-  Future<void> invalidate(Iterable<AssetId> ids) async {
-    // First finish all pending operations, as they will write to the cache.
-    for (var id in ids) {
-      await _canReadCache.remove(id);
-      await _pendingBytesContentCache.remove(id);
-      await _pendingStringContentCache.remove(id);
-    }
+  void invalidate(Iterable<AssetId> ids) {
     for (var id in ids) {
+      _pendingWrites.remove(id);
+      _pendingDeletes.remove(id);
+      _canReadCache.remove(id);
       _bytesContentCache.remove(id);
       _stringContentCache.remove(id);
     }
   }
 
   @override
-  Future<bool> exists(
-    AssetId id, {
-    required Future<bool> Function() ifAbsent,
-  }) => _canReadCache.putIfAbsent(id, ifAbsent);
+  void flush() {
+    for (final write in _pendingWrites.values) {
+      write();
+    }
+    for (final delete in _pendingDeletes.values) {
+      delete();
+    }
+    _pendingWrites.clear();
+    _pendingDeletes.clear();
+  }
 
   @override
-  Future<Uint8List> readAsBytes(
-    AssetId id, {
-    required Future<Uint8List> Function() ifAbsent,
-  }) {
-    var cached = _bytesContentCache[id];
-    if (cached != null) return Future.value(cached);
-
-    return _pendingBytesContentCache.putIfAbsent(id, () async {
-      final result = await ifAbsent();
-      _bytesContentCache[id] = result;
-      unawaited(_pendingBytesContentCache.remove(id));
-      return result;
-    });
+  bool exists(AssetId id, {required bool Function() ifAbsent}) {
+    final maybeResult = _canReadCache[id];
+    if (maybeResult != null) return maybeResult;
+    return _canReadCache[id] = ifAbsent();
```

Comment:
I got confused about the `_canReadCache` name. Maybe `_existsCache` would make more sense?